keps/sig-apps/4017-pod-index-label/README.md

```diff
@@ -0,0 +1,879 @@
+which is somewhat hacky solution. Furthermore, the pod index cannot be utilized by the downward API,
+nor used for use cases such as filtering metrics/logs or routing traffic to a pod with a specific index.
+
+To address these issues, we propose to set the pod index as a **label** on pods for both StatefulSets and
+Indexed Jobs. This will allow the pod index to be utilized via the downward API, so StatefulSet pods
+can easily know their own index. In addition, this will allow metrics and logs to more easily be
+filtered by pod index, for both StatefulSets and Indexed Jobs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Set pod index as label on pods for StatefulSet/Indexed Jobs.
+- Adding the label should not be disruptive to existing workloads.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+At a high level, the proposal is to modify the StatefulSet and Job controllers to set the pod index
+as a pod label at pod creation time (for jobs, this would only apply to jobs in
+Indexed completion mode). The details of this are outlined in the Design Details section below.
+
+StatefulSet pod label: `statefulset.kubernetes.io/pod-index`
+Indexed Job pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+
+#### Story 1
+As a user, I would like to lookup a job's pod logs by its index.
+
+#### Story 2
+As a user, I would like to target traffic to a specific pod index (e.g., index 0) in a StatefulSet.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
```

Comment:
One thing you may want to consider is what kind of validation do you want to perform on existing labels in pods?

I think it shouldn't be expected that users are setting k8 labels but maybe thing about some kind of validation to protect this or warn against it?  Validation is really difficult to make stricter so may be worth thinking about what kind of rules you want to put in place now.